sqlmesh/core/macros.py

```diff
@@ -648,6 +649,228 @@ def eval_(evaluator: MacroEvaluator, condition: exp.Condition) -> t.Any:
+    for field in fields:
+        terms.append(exp.func("COALESCE", field, 0))
+    return case.else_(reduce(lambda a, b: a + b, terms))
+
+
+@macro()
+def safe_sub(_: MacroEvaluator, *fields: exp.Expression) -> exp.Case:
+    """Subtract numbers, substitutes nulls for 0s and only returns null if all fields are null.
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "SELECT @SAFE_SUB(a, b) FROM foo"
+        >>> MacroEvaluator().transform(parse_one(sql)).sql()
+        'SELECT CASE WHEN a IS NULL AND b IS NULL THEN NULL ELSE COALESCE(a, 0) - COALESCE(b, 0) END FROM foo'
+    """
+    null_cond = exp.and_(*[field.is_(exp.null()) for field in fields])
+    case = exp.Case().when(null_cond, exp.null())
+    terms: t.List[exp.Func | exp.Sub] = []
+    for field in fields:
+        terms.append(exp.func("COALESCE", field, 0))
+    return case.else_(reduce(lambda a, b: a - b, terms))
+
+
+@macro()
+def safe_div(_: MacroEvaluator, numerator: exp.Expression, denominator: exp.Expression) -> exp.Div:
+    """Divides numbers, returns null if the denominator is 0.
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "SELECT @SAFE_DIV(a, b) FROM foo"
+        >>> MacroEvaluator().transform(parse_one(sql)).sql()
+        'SELECT a / CASE WHEN b = 0 THEN NULL ELSE b END FROM foo'
+    """
+    return numerator / exp.Case().when(denominator.eq(0), exp.null()).else_(denominator)
+
+
+@macro()
+def union(evaluator: MacroEvaluator, *tables: exp.Table) -> exp.Union:
+    """Returns a UNION of the given tables.
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "@UNION(foo, bar)"
+        >>> MacroEvaluator(schema={"foo": {"a": "int", "b": "string", "c": "string"}, "bar": {"a": "int", "b": "int", "c": "string"}}).transform(parse_one(sql)).sql()
+        'SELECT CAST(a AS INT) AS a, CAST(c AS TEXT) AS c FROM foo UNION SELECT CAST(a AS INT) AS a, CAST(c AS TEXT) AS c FROM bar'
+    """
+    column_sets: t.Dict[str, t.Set[t.Tuple[str, exp.DataType]]] = {}
+    columns_seen: t.Dict[str, None] = OrderedDict()  # Ensure order is deterministic
+    for table in tables:
+        map = evaluator.columns_to_types(table.sql())
+        column_sets[table.sql()] = set(map.items())
+        for c in map:
+            columns_seen[c] = None
+    superset = reduce(lambda a, b: a.intersection(b), column_sets.values())
+    precedence = {c: i for i, c in enumerate(columns_seen.keys())}
+    projection = [
+        exp.cast(exp.column(name), typ).as_(name)
+        for name, typ in sorted(superset, key=lambda c: precedence[c[0]])
+    ]
+    selects: t.List[exp.Unionable] = [exp.select(*projection).from_(t) for t in tables]
+    return t.cast(exp.Union, reduce(lambda a, b: a.union(b), selects))
+
+
+@macro()
+def width_bucket(
```

Comment:
width bucket is already a sql function in many dialects, this shouldn't be here, 

if this isn't in bigquery, it should be in sqlglot, can you remove this one

Reply:
ya sure, would you mind giving a pointer as to how I would PR there? good to keep working between the codebases
maybe on slack l8er

Reply:
you'd create a new expression type widthbucket and handle generation of it in bigquery

Reply:
let's delete this for now